src/main/java/com/emc/ecs/tool/BucketWipeOperations.java

```diff
@@ -0,0 +1,228 @@
+            }
+
+            for (S3Object object : listing.getObjects()) {
+                result.setLastKey(object.getKey());
+                submitTask(new DeleteObjectTask(client, bucket, RestUtil.urlDecode(object.getKey())), result);
+            }
+
+            subPrefixes.addAll(listing.getCommonPrefixes());
+        } while (listing.isTruncated());
+
+        for(String subPrefix : subPrefixes) {
+            deleteAllObjectsHierarchical(bucket, subPrefix, result);
+        }
+
+        result.allActionsSubmitted();
+    }
+
+    /**
+     * Deletes all Keys from the specified bucket that start with the given prefix
+     *
+     * @param bucket the target bucket
+     * @param prefix key prefix of objects to be deleted
+     * @param result the asynchronous result of the operation
+     */
+    public void deleteAllObjects(String bucket, String prefix, BucketWipeResult result) {
+        ListObjectsResult listing = null;
+        ListObjectsRequest request = new ListObjectsRequest(bucket).withPrefix(prefix).withEncodingType(EncodingType.url);
+        do {
+            if (listing == null) listing = client.listObjects(request);
+            else listing = client.listMoreObjects(listing);
+
+            for (S3Object object : listing.getObjects()) {
+                result.setLastKey(object.getKey());
+                submitTask(new DeleteObjectTask(client, bucket, object.getKey()), result);
+            }
+
+        } while (listing.isTruncated());
+
+        result.allActionsSubmitted();
+    }
+
+    /**
+     * Deletes all Object versions from the specified bucket that start with the given prefix
+     *
+     * @param bucket the target bucket
+     * @param prefix key prefix of object versions to be deleted
+     * @param result the asynchronous result of the operation
+     */
+    public void deleteAllVersions(S3Client client, String bucket, String prefix, BucketWipeResult result) {
+        ListVersionsResult listing = null;
+        ListVersionsRequest request = new ListVersionsRequest(bucket).withPrefix(prefix).withEncodingType(EncodingType.url);
+        do {
+            if (listing != null) {
+                request.setKeyMarker(listing.getNextKeyMarker());
+                request.setVersionIdMarker(listing.getNextVersionIdMarker());
+            }
+            listing = client.listVersions(request);
+
+            for (AbstractVersion version : listing.getVersions()) {
+                result.setLastKey(version.getKey() + " (version " + version.getVersionId() + ")");
+                submitTask(new DeleteVersionTask(client, bucket, RestUtil.urlDecode(version.getKey()), version.getVersionId()), result);
+            }
+
+        } while (listing.isTruncated());
+
+        result.allActionsSubmitted();
+    }
+
+    /** Submits a task to be executed recording the fact in the result.  The result is updated as the task completes */
+    private void submitTask(Runnable task, BucketWipeResult result) {
```

Comment:
Previously, there were two mechanisms in place to avoid OOM for very large lists/buckets (billions of objects).  One method was to remove Futures from the results as chunks of tasks completed.  This has been mitigated by using counts instead, but the other method was to limit the task queue by using the EnhancedThreadPoolExecutor.blockingSubmit() and a fixed task queue - that has been removed here.  We definitely need that check or this tool will run out of memory as millions of tasks are added (think of a list file with 2B entries).  Maybe we can use a semaphore in the result object, and wait in actionOutstanding() if that count is above say 10k.  Then in actionComplete() we could notify().